servo/connectors/newrelic.py

```diff
@@ -0,0 +1,340 @@
+    license=License.APACHE2,
+    maturity=Maturity.EXPERIMENTAL,
+)
+class NewrelicConnector(BaseConnector):
+    """NewrelicConnector objects enable servo assemblies to capture
+    measurements from the [Newrelic](https://newrelic.com/) metrics server.
+    """
+    config: NewrelicConfiguration
+
+    # TODO
+    # @on_event()
+    # async def check(self,
+    #     filter_: Optional[Filter] = None, 
+    #     halt_on: HaltOnFailed = HaltOnFailed.requirement
+    # ) -> List[Check]:
+    #     """Checks that the configuration is valid and the connector can capture        
+    #     measurements from Newrelic.
+
+    #     Checks are implemented in the NewrelicChecks class.
+
+    #     Args:
+    #         filter_ (Optional[Filter], optional): A filter for limiting the
+    #             checks that are run. Defaults to None.
+    #         halt_on (HaltOnFailed, optional): When to halt running checks.
+    #             Defaults to HaltOnFailed.requirement.
+
+    #     Returns:
+    #         List[Check]: A list of check objects that report the outcomes of the            
+    #             checks that were run.
+    #     """        
+    #     return await NewrelicChecks.run(self.config, filter_, halt_on=halt_on)
+
+    @on_event()
+    def describe(self) -> Description:
+        """Describes the current state of Metrics measured by querying Newrelic.
+
+        Returns:
+            Description: An object describing the current state of metrics
+                queried from Newrelic.
+        """
+        return Description(metrics=self.config.metrics)
+
+    @property
+    @on_event()
+    def metrics(self) -> List[Metric]:
+        """Returns the list of Metrics measured through Newrelic queries.
+
+        Returns:
+            List[Metric]: The list of metrics to be queried.
+        """
+        return self.config.metrics
+
+    @on_event()
+    def instances(self) -> List[str]:
+        async with self.api_client as client:
+            try:
+                response = await client.get('/config')
+                response.raise_for_status()
+            except (httpx.HTTPError, httpcore._exceptions.ReadTimeout, httpcore._exceptions.ConnectError) as error:                
+                self.logger.trace(f"HTTP error encountered during oco config GET {self.optimizer.api_url}/config: {error}")
+                raise
+
+        opsani_config = response.json()
+        canary = opsani_config.get('adjustment', {}).get('control', {}).get('userdata', {}).get('canary')
+        canary_ip = canary.get('ip')
+        canary_ip = re.sub(r"\.", "-", canary_ip)
+        canary_hostname = f"ip-{canary_ip}.ec2.internal"
+
+        async with httpx.AsyncClient(
+            base_url=self.config.api_url + 'applications/{newrelic_app_id}/instances.json'.format(NEWRELIC_APM_APP_ID), 
+            params={'filter[hostname]': canary_hostname},
+            headers={'X-Api-Key': NEWRELIC_APM_API_KEY},
```

Comment:
`NEWRELIC_APM_API_KEY` must come from config to support multiple New Relic connector instances